### PR TITLE
Add missing nested answers to the questionnaire response item map

### LINKED ISSUE #652: Nested questions behave strangely
Originally reported by @joiskash 

**Describe the bug**
- questionnaireResponseChangedCallback not called in drop-down after selecting an answer
- Clear answer items if enableWhen condition is false. This happens when you select an answer, the conditional logic is true, and then you deselect. Not clearing the answer corrupts the questionnaire response. Because the questionnaire response will now have both answers,even though it should actually have the last edit.
- enableWhen bug questionnaireResponseItem.answer.any instead of questionnaireResponseItem.contains
- Handle nested items in answers for linkIdMapsQuestionnaireResponseMap

**Component**
SDC lib

**To Reproduce**
Use the following questionnaire in the gallery app.

**Expected behavior**
Behaves normally

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: Pixel4a

**Additional context**
questionnaire:
```
{
  "meta": {
    "profile": [
      "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire|2.7"
    ],
    "tag": [
      {
        "code": "lformsVersion: 29.0.3"
      }
    ]
  },
  "version": "1",
  "status": "draft",
  "experimental": true,
  "publisher": "IPRD",
  "name": "Malaria net distribution",
  "title": "Malaria Net Survey",
  "resourceType": "Questionnaire",
  "item": [
    {
      "type": "choice",
      "code": [
        {
          "code": "lga-codes",
          "display": "LGA",
          "system": "https://www.oyo.com/codes"
        }
      ],
      "extension": [
        {
          "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
          "valueCodeableConcept": {
            "coding": [
              {
                "system": "http://hl7.org/fhir/questionnaire-item-control",
                "code": "drop-down",
                "display": "Drop down"
              }
            ],
            "text": "Drop down"
          }
        }
      ],
      "required": true,
      "linkId": "1",
      "text": "LGA",
      "answerOption": [
        {
          "valueCoding": {
            "code": "oluyole",
            "display": "OLUYOLE",
            "system": "https://www.oyo.com/codes/lga"
          }
        },
        {
          "valueCoding": {
            "code": "akinyele",
            "display": "AKINYELE",
            "system": "https://www.oyo.com/codes/lga"
          }
        }
      ],
      "item": [
        {
          "type": "choice",
          "code": [],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
              "valueCodeableConcept": {
                "coding": [
                  {
                    "system": "http://hl7.org/fhir/questionnaire-item-control",
                    "code": "drop-down",
                    "display": "Drop down"
                  }
                ],
                "text": "Drop down"
              }
            }
          ],
          "required": false,
          "linkId": "ward-akinyele",
          "text": "Ward",
          "enableWhen": [
            {
              "answerCoding": {
                "code": "akinyele",
                "system": "https://www.oyo.com/codes/lga"
              },
              "question": "1",
              "operator": "="
            }
          ],
          "answerOption": [
            {
              "valueCoding": {
                "code": "ikereku",
                "display": "Ikereku",
                "system": "https://www.oyo.com/codes/wards"
              }
            },
            {
              "valueCoding": {
                "code": "iroko",
                "display": "Iroko",
                "system": "https://www.oyo.com/codes/wards"
              }
            }
          ],
          "item": [
            {
              "type": "choice",
              "code": [
                {
                  "code": "ward-Iroko",
                  "display": "Net distribution",
                  "system": "https://www.oyo.com/codes"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "coding": [
                      {
                        "system": "http://hl7.org/fhir/questionnaire-item-control",
                        "code": "drop-down",
                        "display": "Drop down"
                      }
                    ],
                    "text": "Drop down"
                  }
                }
              ],
              "required": false,
              "linkId": "ward-Iroko",
              "text": "Net distribution",
              "enableWhen": [
                {
                  "answerCoding": {
                    "code": "iroko",
                    "system": "https://www.oyo.com/codes/wards"
                  },
                  "question": "ward-akinyele",
                  "operator": "="
                },
                {
                  "answerCoding": {
                    "code": "akinyele",
                    "system": "https://www.oyo.com/codes/lga"
                  },
                  "question": "1",
                  "operator": "="
                }
              ],
              "enableBehavior": "all",
              "answerOption": [
                {
                  "valueCoding": {
                    "code": "hla-primary-school",
                    "display": "HLA Primary School",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                },
                {
                  "valueCoding": {
                    "code": "christ-apostolic-church",
                    "display": "Christ Apostolic Church",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                }
              ]
            },
            {
              "type": "choice",
              "code": [
                {
                  "code": "ward-ikereku",
                  "display": "Net distribution",
                  "system": "https://www.oyo.com/codes"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "coding": [
                      {
                        "system": "http://hl7.org/fhir/questionnaire-item-control",
                        "code": "drop-down",
                        "display": "Drop down"
                      }
                    ],
                    "text": "Drop down"
                  }
                }
              ],
              "required": false,
              "linkId": "ward-ikereku",
              "text": "Net distribution",
              "enableWhen": [
                {
                  "answerCoding": {
                    "code": "ikereku",
                    "system": "https://www.oyo.com/codes/wards"
                  },
                  "question": "ward-akinyele",
                  "operator": "="
                }
              ],
              "answerOption": [
                {
                  "valueCoding": {
                    "code": "ogun-state-water-corporation",
                    "display": "Ogun state water corporation Arakanga",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                },
                {
                  "valueCoding": {
                    "code": "iberekodo-central-mosque",
                    "display": "Iberekodo Central Mosque",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "choice",
          "code": [
            {
              "code": "ward-oluyole",
              "display": "Ward oluyole",
              "system": "https://www.oyo.com/codes"
            },
            {
              "code": "ward-codes",
              "display": "Ward",
              "system": "https://www.oyo.com/codes"
            }
          ],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
              "valueCodeableConcept": {
                "coding": [
                  {
                    "system": "http://hl7.org/fhir/questionnaire-item-control",
                    "code": "drop-down",
                    "display": "Drop down"
                  }
                ],
                "text": "Drop down"
              }
            }
          ],
          "required": false,
          "linkId": "ward-oluyole",
          "text": "Ward",
          "enableWhen": [
            {
              "answerCoding": {
                "code": "oluyole",
                "system": "https://www.oyo.com/codes/lga"
              },
              "question": "1",
              "operator": "="
            }
          ],
          "answerOption": [
            {
              "valueCoding": {
                "code": "okanhinde",
                "display": "Okanhinde",
                "system": "https://www.oyo.com/codes/wards"
              }
            },
            {
              "valueCoding": {
                "code": "ayegun",
                "display": "Ayegun",
                "system": "https://www.oyo.com/codes/wards"
              }
            }
          ],
          "item": [
            {
              "type": "choice",
              "code": [
                {
                  "code": "ward-okanhinde",
                  "display": "Net distribution",
                  "system": "https://www.oyo.com/codes"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "coding": [
                      {
                        "system": "http://hl7.org/fhir/questionnaire-item-control",
                        "code": "drop-down",
                        "display": "Drop down"
                      }
                    ],
                    "text": "Drop down"
                  }
                }
              ],
              "required": false,
              "linkId": "ward-okanhinde",
              "text": "Net distribution",
              "enableWhen": [
                {
                  "answerCoding": {
                    "code": "okanhinde",
                    "system": "https://www.oyo.com/codes/wards"
                  },
                  "question": "ward-oluyole",
                  "operator": "="
                },
                {
                  "answerCoding": {
                    "code": "oluyole",
                    "system": "https://www.oyo.com/codes/lga"
                  },
                  "question": "1",
                  "operator": "="
                }
              ],
              "enableBehavior": "all",
              "answerOption": [
                {
                  "valueCoding": {
                    "code": "olojuoro-school",
                    "display": "Olojuoro Pry. Sch.",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                },
                {
                  "valueCoding": {
                    "code": "agbeja-school",
                    "display": "Agbeja N. P. Sch",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                }
              ]
            },
            {
              "type": "choice",
              "code": [
                {
                  "code": "ward-ayegun",
                  "display": "Net distribution",
                  "system": "https://www.oyo.com/codes"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "coding": [
                      {
                        "system": "http://hl7.org/fhir/questionnaire-item-control",
                        "code": "drop-down",
                        "display": "Drop down"
                      }
                    ],
                    "text": "Drop down"
                  }
                },
                {
                  "url": "http://lhcforms.nlm.nih.gov/fhirExt/dataControl",
                  "valueString": "[]"
                }
              ],
              "required": false,
              "linkId": "ward-ayegun",
              "text": "Net distribution",
              "enableWhen": [
                {
                  "answerCoding": {
                    "code": "ayegun",
                    "system": "https://www.oyo.com/codes/wards"
                  },
                  "question": "ward-oluyole",
                  "operator": "="
                }
              ],
              "answerOption": [
                {
                  "valueCoding": {
                    "code": "aspen-arena",
                    "display": "ASPEN ARENA",
                    "system": "https://www.oyo.com/codes/distributionCenter"
                  }
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "type": "integer",
      "code": [
        {
          "code": "number-nets",
          "display": "Number of nets",
          "system": "https://www.oyo.com/codes"
        }
      ],
      "extension": [
        {
          "url": "http://hl7.org/fhir/StructureDefinition/maxValue",
          "valueInteger": 5
        }
      ],
      "required": true,
      "linkId": "number-nets",
      "text": "Number of nets"
    },
    {
      "type": "string",
      "code": [],
      "required": true,
      "linkId": "head-of-household",
      "text": "Initials of the Head of the Household"
    },
    {
      "type": "boolean",
      "code": [
        {
          "code": "phone-number-permission",
          "display": "Does Head of household agree to provide phone number for healthcare use?"
        }
      ],
      "required": true,
      "linkId": "phone-number-permission",
      "text": "Does Head of household agree to provide phone number for healthcare use?",
      "initial": [
        {
          "valueBoolean": "false"
        }
      ],
      "item": [
        {
          "type": "choice",
          "code": [
            {
              "code": "country-code",
              "display": "Country code",
              "system": "https://countrycode.com/countryCodes"
            }
          ],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
              "valueCodeableConcept": {
                "coding": [
                  {
                    "system": "http://hl7.org/fhir/questionnaire-item-control",
                    "code": "drop-down",
                    "display": "Drop down"
                  }
                ],
                "text": "Drop down"
              }
            }
          ],
          "required": false,
          "linkId": "country-code",
          "text": "Country code",
          "enableWhen": [
            {
              "answerBoolean": true,
              "question": "phone-number-permission",
              "operator": "="
            }
          ],
          "enableBehavior": "all",
          "answerOption": [
            {
              "valueCoding": {
                "code": "nigeria",
                "display": "+234",
                "system": "https://countrycode.com/countryCodes"
              }
            },
            {
              "valueCoding": {
                "code": "india",
                "display": "+91",
                "system": "https://countrycode.com/countryCodes"
              }
            },
            {
              "valueCoding": {
                "code": "usa",
                "display": "+1",
                "system": "https://countrycode.com/countryCodes"
              }
            },
            {
              "valueCoding": {
                "code": "south-africa",
                "display": "+27",
                "system": "https://countrycode.com/countryCodes"
              }
            }
          ]
        },
        {
          "type": "string",
          "code": [
            {
              "code": "phone-number",
              "display": "Phone number",
              "system": "https://countrycode.com/countryCodes"
            }
          ],
          "required": false,
          "linkId": "phone-number",
          "text": "Phone number",
          "enableWhen": [
            {
              "answerBoolean": true,
              "question": "country-code",
              "operator": "exists"
            }
          ],
          "enableBehavior": "all"
        }
      ]
    }
  ]
}
```

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #652 

**Description**

- In the questionnaire response item map, we were missing answers to questions nested under other non-group questions. This causes issues for enable when evaluation because no answer can be retrieved to evaluate if another question should be enabled or not
- The assumption that the response item map is immutable needs to be removed. This is because as the user answers questions, more nested items might be created (questions nested under other questions are nested in the response under the answers).

**Alternative(s) considered**
NA

**Type**
Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -80,7 +80,10 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
   private val questionnaireResponseItemChangedCallback: (String) -> Unit = { linkId ->
     linkIdToQuestionnaireItemMap[linkId]?.let {
       if (it.hasNestedItemsWithinAnswers) {
-        linkIdToQuestionnaireResponseItemMap[linkId]!!.addNestedItemsToAnswer(it)
+        linkIdToQuestionnaireResponseItemMap[linkId]?.addNestedItemsToAnswer(it)
+        linkIdToQuestionnaireResponseItemMap[linkId]?.answer?.singleOrNull()?.item?.forEach {
+          linkIdToQuestionnaireResponseItemMap[it.linkId] = it
+        }
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:86]: Optional: Better readability if we specify "it" .. for the outer let `questionnaireItemComponent ->` and inside the forEach `questionnaireResponseItemComponent`

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 2]
```diff
@@ -123,13 +126,16 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
...
-  ): Map<String, QuestionnaireResponse.QuestionnaireResponseItemComponent> {
+  ): MutableMap<String, QuestionnaireResponse.QuestionnaireResponseItemComponent> {
     val linkIdToQuestionnaireResponseItemMap =
       questionnaireResponseItemList.map { it.linkId to it }.toMap().toMutableMap()
     for (item in questionnaireResponseItemList) {
       linkIdToQuestionnaireResponseItemMap.putAll(
         createLinkIdToQuestionnaireResponseItemMap(item.item)
       )
+      item.answer.forEach {
+        createLinkIdToQuestionnaireResponseItemMap(it.item)
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:137]: We need the putAll or else the map will not have them
```
item.answer.forEach {
        linkIdToQuestionnaireResponseItemMap.putAll(
          createLinkIdToQuestionnaireResponseItemMap(it.item)
        )
      }
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 3]
```diff
@@ -80,7 +80,10 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
   private val questionnaireResponseItemChangedCallback: (String) -> Unit = { linkId ->
     linkIdToQuestionnaireItemMap[linkId]?.let {
       if (it.hasNestedItemsWithinAnswers) {
-        linkIdToQuestionnaireResponseItemMap[linkId]!!.addNestedItemsToAnswer(it)
+        linkIdToQuestionnaireResponseItemMap[linkId]?.addNestedItemsToAnswer(it)
```
Comment (stevenckngaa) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:83]: Could linkIdToQuestionnaireResponseItemMap[linkId] be null? Wouldn't line 81 guarantee that linkIdToQuestionnaireResponseItemMap[linkId] isn't null unless there is a potential race condition somewhere?

**Review by joiskash — CHANGES_REQUESTED** [review]
Please look at the comments.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 2 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 3 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Not really - the questionnaire response might not have been created for certain nested items yet. I've simplified the code to address this.

**Comment by jingtang10** [comment]
thanks @joiskash @stevenckngaa please take anther look!

**Review by joiskash — APPROVED** [review]
Awesome! thanks

**Pull request merged**